### DISCUSSION TITLE
Passing arbitrary data to Cast / ValidationRule

### DISCUSSION THREAD

**Discussion by tborealis** [spatie/laravel-data#615] [Q&A]
I have a situation where I need to pass arbitrary data to casts and validation rules. I've mostly looked into the Cast for now. I see that you can pass literal values to casts via the attribute but I need the ability to pass a computed value. I don't want to pollute the payload with this data, mostly due to it coming from a request. I've done the following for now: 

```php
final class AddLibraryRetailerProductsDto extends Data
{
    public static function fromRequestWithCompany(Request $request, Company $company): self
    {
        $payload = $request->all();
        $payload['_options']['company'] = $company;

        return self::from($payload);
    }

    public function __construct(
        #[WithCast(RetailerProductFromUuidCast::class)]
        public readonly RetailerProduct $retailerProduct,
    ) {
    }
}
```

This works but I'm not a fan of using the payload in this way.

For validation rules, I can't see any way to get arbitrary data at the point the rules are constructed. Any suggestions would be appreciated.

**Comment by rubenvanassche** [comment]
If your Cast always will be performed when coming from a request, why not inject the request which contains the company I expect?

Otherwise, use a magic method to create the data object and uuid, these are difficult cases which are a bit harder to get right.

**Reply by tborealis** [reply]
Company in the above example is derived from the logged in user but not a part of the request. I didn't make this very clear in the original question, but I don't like the idea that part of my resource authorisation is mixed with the request, hence not putting it in the payload prior to this point. I've gone with what I have shown above, it seems like the cleanest approach for now.

Any thoughts on passing data into a validation rule that is not part of the payload?
